docs/guides/single-fetch.md

```diff
@@ -4,99 +4,97 @@ title: Single Fetch
 
-```tsx filename=app/entry.server.tsx lines=[1-2,32-33]
-// Reject all pending promises from handler functions after 5 seconds
-export const streamTimeout = 5000;
+## Adding a New Route with Single Fetch
 
-// ...
+With Single Fetch enabled, you can go ahead and author routes that take advantage of the more powerful streaming format and `response` stub
 
-function handleBrowserRequest(
-  request: Request,
-  responseStatusCode: number,
-  responseHeaders: Headers,
-  remixContext: EntryContext
-) {
-  return new Promise((resolve, reject) => {
-    const { pipe, abort } = renderToPipeableStream(
-      <RemixServer
-        context={remixContext}
-        url={request.url}
-        abortDelay={ABORT_DELAY}
-      />,
-      {
-        onShellReady() {
-          /* ... */
-        },
-        onShellError(error: unknown) {
-          /* ... */
-        },
-        onError(error: unknown) {
-          /* ... */
-        },
-      }
-    );
+<docs-info>In order to get proper type inference, you first need to add `@remix-run/react/future/single-fetch.d.ts` to the end of your `tsconfig.json`'s `compilerOptions.types` array. You can read more about this in the [Type Inference section](#type-inference).</docs-info>
 
-    // Automatically timeout the react renderer after 10 seconds
-    setTimeout(abort, 10000);
-  });
+```tsx
+import { unstable_defineLoader as defineLoader } from "@remix-run/node";
+
+export const loader = defineLoader(async () => {
+  const data = await fetchSomeData();
+  return {
+    message: data.message, // <- string
+    date: data.date, // <- Date
+  };
+});
+
+export default function Component() {
+  const data = useLoaderData<typeof loader>();
+  //    ^? { message: string, date: Date }
 }
 ```
 
+## Migrating a Route with Single Fetch
+
+The following changes are not required to take advantage of Single Fetch (refer to [Enabling Single Fetch][start] to get started). We recommend making the following changes on a route-by-route basis, as it's easier to validate changes to things like headers and data types.
+
+Making these changes will ensure a smooth, non-breaking upgrade to React Router v7.
+
 ### Type Inference
 
-Without Single Fetch, any plain Javascript object returned from a `loader` or `action` is automatically serialized into a JSON response (as if you returned it via `json`). The type inference assumes this is the case and infer naked object returns as if they were JSON serialized.
+Without Single Fetch, any plain Javascript object returned from a `loader` or `action` is automatically serialized into a JSON response (as if you returned it via `json`). The type inference assumes this is the case and infers naked object returns as if they were JSON serialized.
```

Comment:
The [Type Inference](https://remix.run/docs/en/main/guides/single-fetch#type-inference) in mentioned here under `Migrating a Route with Single Fetch` section as  optional change. 

But it is also mentioned in `Breaking Changes`
![Screenshot 2024-05-30 at 5 15 24 PM](https://github.com/remix-run/remix/assets/1610169/db2fcaa4-2443-4886-90bb-c7663e4a0eca)

A little confuse as to if this is a Breaking or Optional change.

Reply:
Yeah, I get what you're saying. It's more that the current type inference is buggy if you return naked objects (since `useLoader<typeof loader>()` assumes that the return value is serialized

There's probably a better way to say that, but basically if you want to get the best type safety you need to adopt `defineLoader`/`defineAction`